earth_enterprise/src/common/khxml/khxml.cpp

```diff
@@ -40,50 +41,147 @@ ListElementTagName(const std::string &tagname)
+//
+//  If you are calling XMLPlatformUtils::Initialize() a number of
+//  times, and then follow with XMLPlatformUtils::Terminate() the same
+//  number of times, only the first XMLPlatformUtils::Initialize() will
+//  do the initialization, and only the last XMLPlatformUtils::Terminate()
+//  will clean up the memory. The other calls are ignored.
+//
+//  To ensure all the memory held by the parser are freed, the number of
+//  XMLPlatformUtils::Terminate() calls should match the number
+//  of XMLPlatformUtils::Initialize() calls.
+//  -------------------------------------------------------------------
+    void ReInitXerces(void) throw()
+    {
+        try {
+            XMLPlatformUtils::Terminate();
+        } catch (const XMLException& toCatch) {
+            notify(NFY_FATAL,"Unable to Terminate and ReInitialize Xerces: %s",
+                   FromXMLStr(toCatch.getMessage()).c_str());
+        }
+        try {
+            XMLPlatformUtils::Initialize(/*MiscConfig::Instance().*/initialDOMHeapAllocSize,
+                                         /*MiscConfig::Instance().*/maxDOMHeapAllocSize,
+                                         /*MiscConfig::Instance().*/maxDOMSubAllocationSize);
+        } catch (const XMLException& toCatch) {
+            notify(NFY_FATAL,"Unable to ReInitialize Xerces: %s",
+                   FromXMLStr(toCatch.getMessage()).c_str());
+        }
+       notify(NFY_NOTICE,"successful xerces reinit...");
+    }
+
 };
 
 static khMutexBase xmlLibLock = KH_MUTEX_BASE_INITIALIZER;
+static khMutexBase reInitOuterLock = KH_MUTEX_BASE_INITIALIZER;
+static khMutexBase reInitInnerLock = kH_MUTEX_BASE_RECURSIVE;
 
-void InitializeXMLLibrary(void) throw()
+void InitializeXMLLibrary(bool reinit=false) throw()
 {
   khLockGuard guard(xmlLibLock);
   static UsingXMLGuard XMLLibGuard;
+  if (reinit) XMLLibGuard.ReInitXerces();
+}
+
+// Logic:
+//
+// $ kill -1 <PID>
+//
+// 1. ReInitializeXMLLibrary declared to handle SIGHUP signal in
+//    UsingXMLGuard::UsingXMLGuard
+// 2. ReInitializeXMLLibrary handles this signal and calls
+//    InitializeXMLLibrary, and uses a flag to signify reinit
+// 3. InitializeXMLLibrary has the lock and calls UsingXMLGuard::ReInitXerces
+//    which calls XMLPlatformUtils::Terminate, then Initialize
+//
+// Currently, reads intermittently cause segfaults. This will be handled
+// in the final form by guards on individual parser creation methods within
+// the main xerces wrapper class. Code remains, but commented out to get an idea
+
+void ReInitializeXMLLibrary(int sig) throw()
+{
+    khLockGuard outer(reInitOuterLock);
+    khLockGuard inner(reInitInnerLock);
+    InitializeXMLLibrary(true);
 }
 
 
 DOMDocument *
 CreateEmptyDocument(const std::string &rootTagname) throw()
 {
+  //khLockGuard outer(reInitOuterLock);
+  //{
+  //khLockGuard inner(reInitInnerLock);
   InitializeXMLLibrary();
 
   try {
     DOMImplementation* impl =
-      DOMImplementationRegistry::getDOMImplementation(0);
```

Comment:
I'd like to see the extra whitespace retained here. It's helpful to have the second line indented so that we know it's a continuation of the previous line. Same thing a few lines down.

Reply:
Will do